apps/jupyter/extensions/notebook-runner/src/notebook-runner-state.ts

```diff
@@ -0,0 +1,276 @@
+import { JupyterFrontEnd } from "@jupyterlab/application";
+import { SessionContext } from "@jupyterlab/apputils";
+import {
+  IComm,
+  IKernelConnection,
+} from "@jupyterlab/services/lib/kernel/kernel";
+import { IDocumentManager } from "@jupyterlab/docmanager";
+import { INotebookTracker } from "@jupyterlab/notebook";
+import { KernelMessage } from "@jupyterlab/services";
+import {
+  addKernelListeners,
+  setKernelIsPrepared,
+  waitForKernelToBePrepared,
+} from "./utils";
+import { importBasePackages, installNbConvert, installOtter } from "./packages";
+
+export type CommListener = (
+  comm: IComm,
+  message: KernelMessage.ICommMsgMsg<"shell" | "iopub">,
+) => Promise<void>;
+
+export class NotebookRunnerState {
+  public static commTargetName = "notebook-runner";
+
+  private _resolveOtterSessionContext: (value: SessionContext) => void =
+    () => {};
+
+  public otterSessionContext: Promise<SessionContext> =
+    new Promise<SessionContext>((resolve) => {
+      this._resolveOtterSessionContext = resolve;
+    });
+
+  private commListeners: CommListener[] = [];
+
+  constructor(
+    protected readonly app: JupyterFrontEnd,
+    protected readonly documentManager: IDocumentManager,
+    protected readonly notebookTracker: INotebookTracker,
+    public allowNextNotebookInParallel: boolean = false,
+  ) {
+    this.init();
+  }
+
+  public addCommListener(listener: CommListener): void {
+    this.commListeners.push(listener);
+  }
+
+  public removeCommListener(listener: CommListener): void {
+    this.commListeners = this.commListeners.filter((l) => l !== listener);
+  }
+
+  private async init(): Promise<void> {
+    // eslint-disable-next-line no-async-promise-executor
```

Comment:
I'm not convinced that this `eslint-disable` is desirable. Given how this is used, I'd tend to agree with eslint. It seems that all of the content of the Promise executor could be inlined in the `init()` function, albeit wrapped with a `try{} catch{}`.